Adding a basic concept (closes #589) (#594)
exercism/cpp: concepts/basics/introduction.md

```cpp
int vip_fee(int standard_fee) {
    /*
    vip_fee calculates the vip fee based on the standard_fee.
    */
    int vip_multi{3};
    return standard_fee * vip_multi;
}
```

Or `total_fee`, a function with three parameters and a call to another function.

```cpp
int total_fee(int vips, int adults, int kids) {
    /*
    total_fee calculates the total price for a group of VIP and adult guests with kids.
    Kids get a flat discount on the standard fee.
    VIP guest fees are calculated by calling vip_fee.
    */
    int standard_fee{30};
    int kids_discount{15};
    
    int kids_total_fee = kids  * (standard_fee - kids_discount);
    int vips_total_fee = vips * vip_fee(standard_fee);
    int adult_total_fee = adults * standard_fee;

    return vips_total_fee + adult_total_fee + kids_total_fee;
}
```

Functions in C++ do not return the value of the last statement like in some other languages.
The `return` keyword is required for the code to compile.

### Whitespace

Whitespace is used for formatting source code and includes spaces, tabs, or newlines.
As the compiler ignores unnecessary whitespace, you can use it to structure your code.

> Smart use of whitespace can thus the readability of your code.

Reviewer: ```suggestion
Smart use of whitespace can improve the readability of your code.
```

> There are different formatting standards, but it is not enforced by the compiler.

Reviewer: ```suggestion
There are different formatting standards, but these are all conventional and not enforced by the compiler.
```

> # Introduction

Reviewer 2: I think this is a very solid introduction, I think a mention of `;` could be valuable, since most languages don't have it / require it. I think it should not be that long but more or saying it has to be after each line could be helpful.

Author: You're right. I think I had it in there once and decided on "show don't tell" because there are so many exceptions to any rule I tried to state. If someone has a good sentence, I'm all in.

To set off the C++ journey we are starting with variables, function calls, and comments.

## Comments

Comments come in two flavors: single- and multi-line.
Everything that comes after `//` on the same line is ignored by the compiler.
Multi-line comments are also known as C-style comments. 
They are surrounded by `/*` and `*/`.
Anything that comes between these will be ignored as well.

## Variables

C++ is a typed language.

> All types need to be known at compile time, and so you generally need to state them explicitly.

Reviewer 3: style: maybe delete "so"

A variable's type cannot change.
An integer variable with the name `years` can be declared like this:

```cpp
int years;
```

It is good practice to initialize variables upon declaration.
C++ offers different mechanisms to do so.
The version with the curly braces is more in line with modern C++, but the equal-sign version is also very common.

```cpp
int tomatoes{80};
int potatoes = 40;
```

```exercism/caution
C++ does allow using uninitialized variables.

> Until the variable is deliberately set, it is undefined and might contain anything.

Reviewer 3: suggestion: Until the value is deliberately set you are not allowed to access it.

Reviewer 4: you are allowed, but it is undefined to read it. assigning to it is very well defined.

Reviewer 3: what's the difference between "this is not allowed" and "if you do that you get UB"? Or are we splitting hairs here?

Reviewer 4: I'm hair-splitting on what `variable access` means. Memory write and memory read both are access in my dictionary but only the write is allowed. you are disallowed to read uninitialized memory to get a well formed program.

on the contrary
```cpp
void do_nothing(int&) {}
int* danger = nullptr;
int main() {
    do_nothing(*danger);
}```
does this access contents in the memory? syntactically we just dereferenced a null pointer.

Reviewer: Where I live the law does not forbid squaring the circle. However, should I one day do it, then who knows what will happen.

But I don't think we should burden the student with the difference. What is important to know is that _it is possible to misbehave_ and also that _you shouldn't misbehave, lest bad things happen_.

I don't care for the «undefined behavior» terminology. In C++ land there are multiple flavors of it (though they go by different names), and outside of it sometimes different phenomena are meant, not always easily mapped onto C++ concepts.

Reviewer 4: > What is important to know is that it is possible to misbehave and also that you shouldn't misbehave, lest bad things happen.

This is what we need to communicate!

I hate when teaching materials lie about some concepts because it is too complicated to explain correctly. If we simplify things it has to be explicitly stated that it is just a friendly little analogy that helps the understanding, but not the whole picture.

Reviewer 3: That's why I wrote "value". An uninitialized *variable* has an indeterminate *value*. You are allowed to assign something to that variable but not to access the value. But I see why that can be confusing.

Reviewer 3: I want to be more strict than "it is possible to misbehave". With UB you get worse things than random data, the whole program can do something completely unexpected. That's why I used the informal term "not allowed".

Reviewer: I know better, but am still vulnerable to interpreting «is not allowed» as meaning _some compiler error will keep me from doing it_. Other people are likely to misintrepret it this way as well.

In https://github.com/exercism/cpp/pull/594#discussion_r1172714482 I tried to avoid the overly technical issues.

To avoid used-before-set errors and undefined behaviour it is adviseable to **always initialize**.
```

## Arithmetic Operations

Arithmetic operators like `*`, `+`, or `-` can be part of an expression like `3 * 2` or `tomatoes + potatoes`. 

## Updating Variables

You can reassign variables, as long as they keep their type:

```cpp
tomatoes = tomatoes - 5; // tomatoes is now 75
potatoes = (32 * 2) + 11; // potatoes is now 75 as well
```

## Functions

> Functions have a name, a return type and (possibly empty) parameter list.

Reviewer 2: Is there a writing style in c++? If functions should be written in snake case it could be worth a mention,

Author: Unfortunately there is no single coding style guideline, but [quite a lot of them](https://gist.github.com/derofim/df604f2bf65a506223464e3ffd96a78a#useful-links). Personally I would opt for the [CppCoreGuidelines style](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#S-naming).  Do you thin we should include that somewhere?

Reviewer 2: I mean it depends on, like for this example, is there an agreed-upon standard to use snake case? Or is it alright to use something like camelcase? Since in a language like Ruby, there is more or less a consensus to write methods (functions) in snake case and if it is a consensus in a language community I think it could be worth mentioning.

Reviewer: > Do you thin we should include that somewhere?

You already went a bit into style in the Whitespace section. It could be expanded to explain that many coding style standards exist. I'd argue against recommending adopting any specific one though. (Or perhaps even mentioning any.)

Students who want to know about the various standards can search for them. Students who _need_ to know we cannot help: their situation dictates it for them.

I do not think there is much value in trying to adhere to any one standard on this track. I just don't see the point of doing so. Using your best judgment case by case seems good enough, less effort, and less arbitrary/quirky.

@meatball133 No, C++ is a wild west.

An example of a function named `always_fortyseven` that would always return 47 would look like this:

```cpp
int always_fortyseven() {
    return 47;
}
```

Here is `vip_fee`, which has one parameter:
```cpp
int vip_fee(int standard_fee) {
    /*
    vip_fee calculates the vip fee based on the standard_fee.
    */
    int vip_multi{3};
    return standard_fee * vip_multi;
}
```

> Or `total_fee`, a function with three parameters and a call to another function.

Reviewer 2: I don't know if this many examples of functions are needed, my suggestion would be to perhaps be to reduce to 2 examples.

# Introduction

To start the C++ journey we are starting with variables, function calls, and comments.

## Variables

C++ is a typed language.

> The type of a variable has to be stated upon the declaration of a variable.

Reviewer: [Lies to children](https://en.wikipedia.org/wiki/Lie-to-children) can be OK. Still, an attempt:

> All types need to be known at compile time, and so you generally need to state them explicitly.

> Its type cannot change during its lifetime.

Author: Should I drop the rest after "Its type cannot change"? It would still be correct, but I would avoid to explain scopes and lifetimes and memory management? @siebenschlaefer

> Its type cannot change during its lifetime.

Reviewer: Lifetimes probably deserve their own concept node. I'm not sure what to do with scopes, but I am still attracted to postponing their discussion.

The statement that variables' types cannot change only makes sense in the context of other languages in which something like that _seems_ to be possible (in dynamic languages _values_ may have types; in static languages _variables_ have types). Therefore perhaps put this discussion in an `exercism/note` block?

An integer variable with the name `years` can be declared like this:

```cpp
int years;
```

> It is good practice to initialize variables upon declaration.

Author: I think I should put that into the numbers concept. If you do not initialize a variable, it will be some kind of null-value. Which is bad form, but okay for an int type that will have "zero" for most compilers but might be garbage and cause unforeseen behaviour for reference types with uninitialized memory blocks.

> It is good practice to initialize variables upon declaration.

Author: I added `This ensures that the content of the variable is well-defined and not filled with random data.`. Is this good reason enough?

> It is good practice to initialize variables upon declaration.

Reviewer: ```exercism/caution
    C++ does allow using uninitialized variables.
    However, when you do the results can be unpredicable.
    [optionally offer examples: an undesirable situation, and a disastrous one]
    It is therefore good practice to initialize variables at their declaration whenever possible.
    That way it is impossible to forget to on time.
    ```

A disadvantage of «content» and «filled» and «random data» is that it might not be clear to a student what these should mean.

> It is good practice to initialize variables upon declaration.

Reviewer: Begs the question what «valid» means. Or what sorts of «invalid» data might even exist.

Coming from many other languages, the idea of uninitialized variables is foreign and _weird_.

C++ offers different mechanisms to do so.
The version with the curly braces is more in line with modern C++, but the equal-sign version is also very common.

```cpp
int tomatoes{80};
int potatoes = 40;
```

## Arithmetic Operations

Arithmetic expressions like `*`, `+`, or `-` can be part of a statement. 

```cpp

> 3 * 2;

Reviewer 3: Here I would omit the semicolon because this is not a complete statement.

// => 6
```

## Updating Variables

> You can reassign variables, as long as keep their type:

Author: I could use "change" but, that would not really cover what happens in the background. Maybe we should do a concept on lifetimes and memory management?

```cpp
tomatoes = tomatoes - 5; // tomatoes is now 75
potatoes = (32 * 2) + 11; // potatoes is now 75 as well
```

> ## Functions

Reviewer 5: It doesn't seem like a lot to me to be fair, it's really a list of essential things to know about C++ functions to be able to write them. I wouldn't add anything else on it though.

> ## Functions

Author: I would rather stick to "Just give the user enough to solve the current exercise." Do they need to know what a statement is or that `return` is a special statement or how evaluations work? Because there are many exceptions to "easy rules" in C++.

Functions have a name and a return type.

> It is possible to pass a parameter list to the function.

Reviewer 3: "pass a parameter list to a function" sounds odd.
How about:
"Functions have a name, a return type and (possibly empty) parameter list."

An example of a function named `always_fortyseven` that would always return 47 would look like this:

```cpp
int always_fortyseven() {
    return 47;
}
```

Or `total_fee`, a function with three parameters.

```cpp
int total_fee(int vip, int adult, int kids) {
    /*
    total_fee calculates the total price for a group of VIP and adult guests with kids.
    Kids get a flat discount on the standard fee.
    VIP guest fees are calculated by multiplying the fee.
    */
    int standard_fee{30};
    int kids_discount{15};
    int vip_multi{3};
    int kids_total_fee = kids  * (standard_fee - kids_discount);

    return vip * vip_multi + standard_fee * adult + kids_total_fee;
}
```

> C++ does not return the last statement in a function like some other languages.

Reviewer: ```suggestion
Functions in C++ do not return the value of the last statement like in some other languages.
```

Reviewer 4: Is this phenomenon so common that it should be stated here?

Author: I have seen it quite a few times, that people ask why their code is not compiling.

> The `return` keyword is required for the code to compile.

Reviewer 3: Could this be misunderstood that `return;` is necessary for `void` functions?

Reviewer: - No: the student does not yet know about `void`, and has only ever seen `return expression;`.
- Yes.

But I don't think it matters here. `void` functions will be dealt with elsewhere in the syllabus, and the student does not need to know about them right now.

Reviewer 4: also `main` can omit the return statement although it is declared as int (without warnings)

Author: That is a technical detail, that I considered out of scope. I think it should be remarked in the function concept.

# Introduction

To set off the C++ journey we are starting with variables, function calls, and comments.

## Comments

Comments come in two flavors: single- and multi-line.

> Everything that comes after `//` in a line is ignored by the compiler.

Reviewer: ```suggestion
Everything that comes after `//` on the same line is ignored by the compiler.
```

Multi-line comments are also known as C-style comments. 

> They are surrounded by `/*` and `*/`.

Reviewer: ~~~suggestion
They are surrounded by `/*` and `*/`.
Anything that comes between these will be ignored as well.

```cpp
code; // these words will be ignored
these_words_will_not_be_ignored; /*
but these
   will be */
```
~~~

Or something like that.

Author: I'm not sure I can whip up a good example, with the comments being the first item that is explained.

int potatoes = 40;
```

```exercism/caution
C++ does allow using uninitialized variables.
Until the variable is deliberately set, it is undefined and might contain anything.

> To avoid used-before-set errors and undefined behaviour it is adviseable to **always initialize**.

Reviewer 4: the word `behaviour` is the British spelling, and the standard uses the American spelling `behavior`.
https://en.cppreference.com/w/cpp/language/ub

}
```

> Here is `vip_fee`, which has one parameter:

Reviewer 4: are we calling them parameters or arguments?

Reviewer: When a function has parameters, you can pass it arguments.
